File: leetcode/230. Kth Smallest Element in a BST.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    int kthSmallest(TreeNode* root, int k) {
        return inorder(root, k);
    }
    
private:
    int inorder(TreeNode* node, int& k) {
        if(node == NULL){
            return 0;
        }
        int left_res = inorder(node->left, k);
        if(k == 0){
            return left_res;
        }
        k--;
        if(k == 0){
            return node->val;
        }
        int right_res = inorder(node->right, k);
        return right_res;
    }
};

File: leetcode/230. Kth Smallest Element in a BST.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def kthSmallest(self, root, k):
        """
        :type root: TreeNode
        :type k: int
        :rtype: int
        """
        cnt = 0
        node =root
        stack = []
        
        while node is not None or len(stack) != 0:
            while node is not None:
                stack.append(node)
                node = node.left
            if len(stack) != 0:
                node = stack.pop()
                cnt += 1
                if cnt == k:
                    return node.val
                node = node.right
